feat(ext/http): UpgradeWebSocketOptions headers

e3dio opened the pull request:
fix https://github.com/denoland/deno/issues/19277

Allow adding headers on Response to Websocket upgrade request

Works for both `serveHttp` and `Deno.serve`

```js
const headers = new Headers({ 'Set-Cookie': 'Hello=123' });
const { socket, response } = Deno.upgradeWebSocket(request, { headers });
```

iuioiua commented:
Why not just modify `response.headers`?

e3dio commented:
@iuioiua Currently the Response headers for websocket upgrade are created "Immutable" and can't be modified: https://github.com/denoland/deno/blob/main/ext/http/01_http.js#L470

This PR adds headers before it is created as immutable. You could instead make the response headers for websocket upgrade mutable. I would be fine with that, I think I would prefer it. Is anyone against making the websocket upgrade response headers mutable, or do people prefer the current PR ? If no one responds I may update this PR to make the headers mutable.

e3dio commented:
@iuioiua I tried setting the headers to mutable and appending to them after generating upgrade response, but it doesn't work, they are doing something strange with the response and not looking at those headers for the actual response. For now going with current PR

iuioiua commented:
I think you could also do:
```ts
const response2 = new Response(response1.body, {
  headers: {
    ...response1.headers,
   "x-custom-header": "foo"
  }
});
```

e3dio commented:
@iuioiua  Nope you did not read my previous post, in the new `Deno.serve` the upgrade and response takes place during the `Deno.upgradeWebSocket` call and not after. The returned response object is simply a placeholder and does nothing. You can read the code to see this

e3dio commented:
The new Deno.serve request has special function set for upgrading websocket here https://github.com/denoland/deno/blob/main/ext/http/01_http.js#L466C13-L466C26

In that function the upgrade and response happens here with the headers that are currently set https://github.com/denoland/deno/blob/main/ext/http/00_serve.js#L195

Then it returns the placeholder response object https://github.com/denoland/deno/blob/main/ext/http/00_serve.js#L228
modifiying this placeholder response or trying to add headers here will do nothing, response as already been sent, as described here https://github.com/denoland/deno/blob/main/ext/http/00_serve.js#L481

e3dio commented:
I'm not saying this is the way it should be, just the way it is currently, and I want to change the least amount of code to get this feature going so I can set headers on my websocket requests ;)

iuioiua commented:
Oh, I see now. Apologies. Yes, I didn't read your comments properly 😆

mmastrac commented:
This looks good, but I think we should consider limiting the headers applied here to an allow-list. I am not sure we need to support anything beyond `Set-Cookie` initially. Are there any others?

Also, we definitely need to have a test in `serve_test.ts`!

e3dio commented:
Personally I only need `Set-Cookie` and am fine with whatever it takes to get that launched, but it seems like you should be able to add whatever headers you want to a response, this is an HTTP server for developers ? But whatever Deno people want to do is fine, I can add a `Deno.serve` test, can someone confirm how they want this to work so this can move forward, and also feel free to update this PR however you want

e3dio commented:
I just read through the WebSocket RFC protocol for opening handshake https://datatracker.ietf.org/doc/html/rfc6455#section-1.3 and it mentions the 3 required headers `upgrade`, `connection`, `sec-websocket-accept` and the optional `sec-websocket-protocol` and then the only other field it mentions is `set-cookie`:
> The server can also set cookie-related option fields to set cookies

So I think I'd be fine with providing the option of only setting cookie header:
```js
const cookies = [ 'Hello=123', 'Cookie2=456' ];
const { socket, response } = Deno.upgradeWebSocket(request, { cookies });
```

mmastrac commented:
I think it makes sense to allow-list one and only one header: `Set-Cookie`. While we don't have any other valid headers right now, I worry that hard-coding cookies as the only valid header will cause us issues in the future if we discover another valid header.

e3dio commented:
So you think do this:

```js
const headers = new Headers({ 'Set-Cookie': 'Hello=123' });
const { socket, response } = Deno.upgradeWebSocket(request, { headers });
```
But in the docs say: "The only allowed header is Set-Cookie" ?

Or just let users set whatever headers they want ? I can't think of a reason to deny a developer a reason to set any header. I don't have a preference I just need Set-Cookie

iuioiua commented:
What about a deny-list instead? It prevents web socket functionality from being tampered with but allows for using any other headers.

mmastrac commented:
A denylist would also work for me.

e3dio commented:
You could do a deny list of the 4 headers that Deno already covers for the response: `upgrade`, `connection`, `sec-websocket-accept` and `sec-websocket-protocol`

That would help developers who might not understand those are already covered by Deno by throwing an error if they try to set them

e3dio commented:
Ok I updated PR, added test for Deno.serve, includes deny header list. I don't like deny header list because it is unnecessary overhead converting strings to lowercase and comparing headers with list for every request. Minor overhead but unnecessary. I don't see why Deno would deny a header from being written. Thinking remove that part

mmastrac reviewed (approved):
LGTM!

e3dio commented:
The test is failing with this error:
![image](https://github.com/denoland/deno/assets/85405955/d02f5e49-be48-4d5a-937e-37987682d683)

But that `forEach` function is not the standard `Array.prototype.forEach`, it's specific to [Headers](https://developer.mozilla.org/en-US/docs/Web/API/Headers/forEach) so I don't think the "Primordial" forEach is right here, does this test need to be manually passed over for this ?

mmastrac reviewed (changes requested):
I took a closer look as we get closer to landing this and I think we need to tweak it a bit. Sorry about jumping the gun and approving too early.

 - First, since this is new API, it should go through a brief unstable period so we need to add it to `lib.deno.unstable.d.ts` and ensure that the correct unstable flag is passed
 - Second, we will need this header data to go through a validation step -- `headersFromHeaderList` should be sufficient

I'll see what I can do to help get this landed as it's something I'd like for the websocket API as well.

bartlomieju commented:
> I took a closer look as we get closer to landing this and I think we need to tweak it a bit. Sorry about jumping the gun and approving too early.
> 
> * First, since this is new API, it should go through a brief unstable period so we need to add it to `lib.deno.unstable.d.ts` and ensure that the correct unstable flag is passed
> * Second, we will need this header data to go through a validation step -- `headersFromHeaderList` should be sufficient
> 
> I'll see what I can do to help get this landed as it's something I'd like for the websocket API as well.

@e3dio would you be able to apply these changes before we release v1.41 tomorrow (Feb, 21st)?

jasonford commented:
Hey @mmastrac and team, I'm also interested in this one landing. It looks like @e3dio is on vacation (thanks for the implementation btw, very timely for us! :). I can devote time getting everything set up to contribute your requested changes next week if that will help most; Though it does look like a relatively light lift for someone who is already comfortable with the repo.

In any case, I was wondering if there is a chance it lands on the next release? I'm very happy to help out next week, or possibly later tomorrow. I'm guessing we might be getting close to a release, looking at the cadence.

jasonford commented:
Hey @mmastrac, I think the PR I submitted to @e3dio's fork (https://github.com/e3dio/deno/pull/1) has what you wanted for 1/2 of your change requests. I'm happy to do the second part, but I wasn't able to see an example of adding a single option in an interface to ```lib.deno.unstable.d.ts```. Seeing as I'm pretty new to developing on this repo, I wasn't sure how to modify things in that file and understand the impact on unstable flag requirements/generated documentation/etc. If you could point me in the right direction or want to make the change yourself, I would very much appreciate it!

jasonford commented:
Hey @bartlomieju Is there a chance this can get added to a patch release for 1.41, or does it need to wait to 1.42? I'm happy to do the second half of @mmastrac's recommendations as well (the first half is in a PR to the fork for this PR: https://github.com/e3dio/deno/pull/1), but I would need a little help understanding how to mark a single option in an interface as unstable. I think that's the only thing blocking this now, and this feature would be super useful to my team. Thanks for all your work!

jasonford commented:
Hey @bartlomieju, I just wanted to see if this has a chance of making it into a patch or perhaps into the 1.42 milestone? Sorry for the repeat messages, but just wanted to make sure this one doesn't get lost in the shuffle since it seems like it is so close and it's a generally useful feature. Thanks again for maintaining such a great tool!

ry commented:
closing this because it's old
